File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int size = nums.length;
        int[] prefixProd = new int[size];
        int[] suffixProd = new int[size];
        int[] res = new int[size];
        prefixProd[0] = 1;
        suffixProd[size-1] = 1;
        for(int i=1;i<size;i++){
            prefixProd[i] = nums[i-1] * prefixProd[i-1];
        } 
        for(int j=size-2;j>=0;j--){
            suffixProd[j] = nums[j+1] * suffixProd[j+1];
        }
        for(int i=0;i<size;i++){
            res[i] = prefixProd[i]*suffixProd[i];
        }
        return res;
    }
}  

File: Data Structures & Algorithms/daily-temperatures/submission-1.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& nums) {
        // How many days do I have to wait until a warmer temperature?
        // we use a stack to remember the indices whose answer we haven't found yet.
        int n = nums.size();
        vector<int> res(n,0);
        stack<int> st;
        for(int i=0;i<n;i++){
            while(!st.empty() && nums[i]>nums[st.top()]){
                int prevIdx = st.top();
                st.pop();
                res[prevIdx] = i-prevIdx;
            }
            st.push(i);
        }
        return res;
    }
};
